src/transform.rs

```diff
@@ -0,0 +1,365 @@
+//! This module implements some transforms of the JSON schema that preserve semantics
+//! but make the schema easier to analyze.
+//!
+//! It turns out that `schemars`'s representation of schemas is pretty clunky for this
+//! purpose. Probably we want a better intermediate representation.
+
+use schemars::schema::{
+    ArrayValidation, ObjectValidation, RootSchema, Schema, SchemaObject, SingleOrVec,
+    SubschemaValidation,
+};
+
+use crate::{references, utils::plain_schema_types};
+
+/// Some JSON schemas factor out common parts of schemas into definitions.
+/// This causes some difficulty for our analysis because information is represented
+/// in multiple different places, so this transformation tries to merge factored-out
+/// parts into a single schema.
+///
+/// As a motivating example, the github-workflow schema factors out
+///
+/// ```json
+///    "eventObject": {
+///      "oneOf": [
+///        {
+///          "type": "object"
+///        },
+///        {
+///          "type": "null"
+///        }
+///      ],
+///      "additionalProperties": true
+///    },
+/// ```
+///
+/// as a definition, and then references it in various places that might instead
+/// contain
+///
+/// ```json
+/// "type": ["object", "null"],
+/// "additionalProperties": true
+/// ```
```

Comment:
I think I don't understand precisely what's going on here. Which one is the factored version? How is the definition used? I would expect that the various places of use would have a `ref` somewhere?

Reply:
I rewrote this comment so hopefully it's clearer.

Reply:
Thanks, it is :+1: additional nitpick (so feel free to ignore): maybe it should be called `inline_def` or `expand_def` instead of `merge_def`?